lib/validator/validator.go

```diff
@@ -129,43 +138,43 @@ func validateLabels(v *validator.Validate, topStruct reflect.Value, currentStruc
 
 func validateInterface(v *validator.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
 	b := []byte(field.String())
-	log.Infof("Validate interface: %s", b)
+	log.Debugf("Validate interface: %s", b)
 	return nameRegex.Match(b)
 }
 
-func validateOrder(v *validator.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
-	f := field.Interface()
-	log.Infof("Validate order: %v", f)
-	return f != nil
-}
-
 func validateASNum(v *validator.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
 	f := field.Interface().(int)
-	log.Infof("Validate AS number: %v", f)
+	log.Debugf("Validate AS number: %v", f)
 	return f >= 0 && f <= 4294967295
 }
 
 func validateScopeGlobalOrNode(v *validator.Validate, topStruct reflect.Value, currentStructOrField reflect.Value, field reflect.Value, fieldType reflect.Type, fieldKind reflect.Kind, param string) bool {
 	f := field.Interface().(scope.Scope)
-	log.Infof("Validate scope: %v", f)
+	log.Debugf("Validate scope: %v", f)
 	return f == scope.Global || f == scope.Node
 }
 
 func validateProtocol(v *validator.Validate, structLevel *validator.StructLevel) {
-	log.Infof("Validate protocol")
 	p := structLevel.CurrentStruct.Interface().(numorstring.Protocol)
-	log.Infof("Validate protocol: %v %s %v", p.Type, p.StrVal, p.NumVal)
-	if p.Type == numorstring.NumOrStringNum && ((p.NumVal < 1) || (p.NumVal > 255)) {
-		structLevel.ReportError(reflect.ValueOf(p.NumVal), "Protocol", "protocol", "protocol number invalid")
-	} else if p.Type == numorstring.NumOrStringString && !protocolRegex.MatchString(p.StrVal) {
-		structLevel.ReportError(reflect.ValueOf(p.StrVal), "Protocol", "protocol", "protocol name invalid")
+	log.Debugf("Validate protocol: %v %s %d", p.Type, p.StrVal, p.NumVal)
+
+	// The protocol field may be an integer 1-254, or one of the valid protocol names.
+	if num, err := p.NumValue(); err == nil {
+		if (num < 1) || (num > 255) {
+			structLevel.ReportError(reflect.ValueOf(p.NumVal), "Protocol", "protocol", "protocol number invalid")
+		}
+	} else if !protocolRegex.MatchString(p.String()) {
```

Comment:
Yup.  You can sort of work around it by setting the String field as "123" but provided you use the "NumValue()" to get the number you should be golden.

When reading from JSON we convert numerical string values to the number.